src/giskardpy/plugin_world_visualization.py

```diff
@@ -0,0 +1,118 @@
+import hashlib
+
+import py_trees
+import rospy
+from visualization_msgs.msg import Marker, MarkerArray
+
+from giskardpy.plugin import GiskardBehavior
+import giskardpy.identifier as identifier
+from giskardpy.tfwrapper import lookup_pose, get_full_frame_name, pose_to_transform, \
+    multiply_transforms, transform_to_pose, list_to_pose
+
+
+class WorldVisualizationBehavior(GiskardBehavior):
+    def __init__(self, name, ensure_publish=False):
+        super(WorldVisualizationBehavior, self).__init__(name)
+        self.map_frame = self.get_god_map().get_data(identifier.map_frame)
+        self.marker_namespace = u'planning_world_visualization'
+        self.ensure_publish = ensure_publish
+        self.currently_publishing_object_names = set()
+        self.links_full_frame_name = {}
+        self.links_visual_offset = {}
+
+    def setup(self, timeout):
+        self.publisher = rospy.Publisher(u'~visualization_marker_array', MarkerArray, queue_size=1)
+        self.ids = set()
+        self.set_full_link_names_for_objects()
+        self.set_links_visual_offsets()
+        return super(WorldVisualizationBehavior, self).setup(timeout)
+
+    def set_full_link_names_for_objects(self):
+        """
+        Resets the full link names (namespace name + link name) for every current
+        object in the world and saves the object names of these objects.
+        """
+        self.currently_publishing_object_names = set()
+        self.links_full_frame_name = {}
+        objects_dict = self.get_world().get_objects()
+        for object_name, object in objects_dict.items():
+            self.currently_publishing_object_names.add(object_name)
+            for link_name in object.get_link_names():
+                if object.has_link_visuals(link_name):
+                    self.links_full_frame_name[link_name] = get_full_frame_name(link_name)
+
+    def set_links_visual_offsets(self):
+        """
+        Saves if a visual offset needs to be calculated for the visuals of every link
+        by saving the offset if the offset is bigger than zero.
+        """
+        self.links_visual_offset = {}
+        objects_dict = self.get_world().get_objects()
+        for object_name, object in objects_dict.items():
+            self.currently_publishing_object_names.add(object_name)
+            for link_name in object.get_link_names():
+                if object.has_link_visuals(link_name) and \
+                        object.get_urdf_link(link_name).visual.origin:
+                    origin = object.get_urdf_link(link_name).visual.origin
+                    if origin.rpy != [0, 0, 0] or origin.xyz != [0, 0, 0]:
+                        id_str = '{}{}'.format(object_name, link_name).encode('utf-8')
+                        self.links_visual_offset[id_str] = list_to_pose([origin.xyz, origin.rpy],
+                                                                        list_euler=True)
```

Comment:
It would be better if `link_as_marker` of `URDFObject` also fills the pose part of the message, instead of doing it in this preprosessing step. I'm not sure why I'm not already doing that. please change the normal visualization plugin accordingly.